### PR TITLE
LowPU cards + highPU recoil implementation on wlike

### PR DESCRIPTION
**Pull request by jeyserma**

### CODE REVIEW

File: scripts/histmakers/mz_wlike_with_mu_eta_pt.py [hunk 1]
```diff
@@ -64,7 +66,7 @@
 
 
 # extra axes for dilepton validation plots
-axis_mll = hist.axis.Regular(24, 60., 120., name = "mll")
+axis_mll = hist.axis.Regular(24, 60., 120., name = "mll", underflow=False, overflow=False)
```
Comment (kdlong) [inline scripts/histmakers/mz_wlike_with_mu_eta_pt.py:69]: Why turn off overflow?

File: wremnants/datasets/datagroupsLowPU.py [hunk 2]
```diff
@@ -20,81 +17,86 @@ def __init__(self, infile, combine=False, flavor=""):
...
+                    color=ROOT.kBlack,
+                    signalOp = None,
+                ),
+            )
+        if flavor == "ee":
+            self.groups.update(
+                SingleElectron=dict(
+                    members=[self.datasets["singleelectron"]],
+                    label="Data",
+                    color=ROOT.kBlack,
```
Comment (kdlong) [inline wremnants/datasets/datagroupsLowPU.py:96]: Maybe you could use the hex color here and in ROOT so that it works with both ROOT and matplotlib

File: wremnants/datasets/datagroupsLowPU.py [hunk 3]
```diff
@@ -20,81 +17,86 @@ def __init__(self, infile, combine=False, flavor=""):
...
                 color=ROOT.TColor.GetColor(100, 192, 232),
-                signalOp = signalOp,
+                signalOp = None,
             ),
             ZZ=dict(
                 members = [self.datasets[x] for x in ["ZZ"]],
                 label="ZZ",
                 color=ROOT.TColor.GetColor(100, 192, 232),
-                signalOp = signalOp,
+                signalOp = None,
```
Comment (kdlong) [inline wremnants/datasets/datagroupsLowPU.py:76]: If you just don't want to select the signal region there is an option to not apply this function when reading the data groups

File: wremnants/include/lowpu_utils.h [hunk 4]
```diff
@@ -162,6 +162,68 @@ Vec_f trigFilter(Vec_f tag, Vec_f trigMatch) {
...
+}
+
+
+// for Z
+const double MZ_GEN_ = 91153.509740726733;
+const double GAMMAZ_GEN_ = 2493.2018986110700;
+const double MW_GEN_ = 80351.812293789408;
+const double GAMMAW_GEN_ = 2090.4310808144846;
+
+double computeBreitWignerWeight(double massVgen, double offset, int type) {
```
Comment (kdlong) [inline wremnants/include/lowpu_utils.h:184]: I think this could reasonably go in the theoryTools.h file as it is of general use. We will probably want to check some things in the high pu using these weights as well.

File: wremnants/theory_tools.py [hunk 5]
```diff
@@ -127,12 +127,12 @@ def define_prefsr_vars(df):
 def define_scale_tensor(df):
     # convert vector of scale weights to 3x3 tensor and clip weights to |weight|<10.
     df = df.Define("scaleWeights_tensor", "wrem::makeScaleTensor(LHEScaleWeight, 10.);")
-    df = df.Define("scaleWeights_tensor_wnom", "auto res = scaleWeights_tensor; res = nominal_weight*res; return res;")
+    df = df.Define("scaleWeights_tensor_wnom", "auto res = scaleWeights_tensor; res = weight*res; return res;")
```
Comment (kdlong) [inline wremnants/theory_tools.py:130]: I think this might be an accidental change, it could cause problems.

File: wremnants/theory_tools.py [hunk 6]
```diff
@@ -155,15 +155,14 @@ def define_and_make_pdf_hists(df, axes, cols, dataset, pdfset="nnpdf31", storeUn
...
     df = df.Define(tensorName, f"auto res = wrem::clip_tensor(wrem::vec_to_tensor_t<double, {entries}>({pdfBranch}), 10.); res = nominal_weight/nominal_pdf_cen*res; return res;")
 
     pdfHist= df.HistoBoost(pdfName, axes, [*cols, tensorName])
 
-    df = df.Define(tensorASName, "Eigen::TensorFixedSize<double, Eigen::Sizes<2>> res; "
-            f"res(0) = {pdfInfo['alphas'][0]}; "
-            f"res(1) = {pdfInfo['alphas'][1]}; "
-            "return wrem::clip_tensor(res, 10.)")
-    alphaSHist = df.HistoBoost(f"alphaS002{pdfName}", axes, [*cols, tensorASName])
+    # slice 2 elements starting from 101
```
Comment (kdlong) [inline wremnants/theory_tools.py:163]: Don't think this is an intended change either. Probably best to make the change to include a name for the pdf hist on top of the current main (could be a separate PR)

File: scripts/histmakers/mz_wlike_with_mu_eta_pt.py (hunk 1 above)
Comment (jeyserma) [reply scripts/histmakers/mz_wlike_with_mu_eta_pt.py:69]: It was for debugging purposes... will revert it.

File: wremnants/datasets/datagroupsLowPU.py (hunk 3 above)
Comment (jeyserma) [reply wremnants/datasets/datagroupsLowPU.py:76]: The signalOp is not defined (not needed for the lowPU so far)

File: wremnants/include/lowpu_utils.h (hunk 4 above)
Comment (jeyserma) [reply wremnants/include/lowpu_utils.h:184]: Moved to utils.h

**Comment by jeyserma** [comment]
@cippy @kdlong I re-arranged how the systematic groups are treated (after the implementation of Marco regarding the suppression of wrong-charged nuisances): a separate function deals now whether or not to add a systematic to a group. It is shared for shape and lnN systematics.
